url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/tools/GameViewer.java

```diff
@@ -0,0 +1,301 @@
+import javax.swing.border.EtchedBorder;
+import javax.swing.border.TitledBorder;
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.awt.event.ItemEvent;
+import java.awt.event.ItemListener;
+
+/**
+ * Created by IntelliJ IDEA.
+ * User: Sergey Karakovskiy
+ * Date: Mar 29, 2009
+ * Time: 3:34:13 PM
+ * Package: .Tools
+ */
+
+public class GameViewer extends JFrame
+{
+    Dimension defaultSize = new Dimension(900, 800);
+    Point defaultLocation = new Point(350, 10);
+
+    //    Thread animator;
+    //    int frame;
+    int delay;
+    int FPS = 5;
+    private MarioVisualComponent marioVisualComponent;
+
+    public void AdjustFPS()
+    {
+        int fps = FPS; // GlobalOptions.FPS;
+        delay = (fps > 0) ? (fps >= GlobalOptions.MaxFPS) ? 0 : (1000 / fps) : 100;
+        System.out.println("Game Viewer animator delay: " + delay);
+    }
+
+    GameViewerView gameViewerViewPanel = new GameViewerView();
+
+    public void setMarioVisualComponent(MarioVisualComponent marioVisualComponent)
+    {
+        this.marioVisualComponent = marioVisualComponent;
+    }
+
+    private class GameViewerView extends JPanel implements Runnable
+    {
+        Thread animator;
+//        private MarioVisualComponent marioVisualComponent;
+
+        public void start()
+        {
+            animator = new Thread(this);
+            animator.start();
+        }
+
+        public void stop()
+        {
+            animator = null;
+        }
+
+        public void paint(Graphics g)
+        {
+            g.fillRect(0, 0, getWidth(), getHeight());
+            g.setColor(Color.YELLOW);
+            int y_dump = 0;
+            g.drawString("Current GAME STATE: ", 320, y_dump += 11 );
+            g.setColor(Color.GREEN);
+            if (this.getMarioVisualComponent() != null)
+            {
+                for (String s: this.getMarioVisualComponent().getTextObservation(
+                        ShowEnemiesObservation.getState(),
+                        ShowLevelMapObservation.getState(),
+                        ShowMergedObservation.getState(),
+                        ZLevelMapValue,
+                        ZLevelEnemiesValue) )
+                {
+                    g.setColor((s.charAt(0) == '~') ? Color.YELLOW : Color.GREEN);
+                    g.drawString(s, 0, y_dump += 11);
+                }
+
+            }
+
+        }
+        
+        public void run()
+        {
+            // Remember the starting time
+            long tm = System.currentTimeMillis();
+            while (Thread.currentThread() == animator)
+            {
+                // Display the next frame of animation.
+                repaint();
+
+                // Delay depending on how far we are behind.
+                try {
+                    tm += delay;
+                    Thread.sleep(Math.max(0, tm - System.currentTimeMillis()));
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)